### ISSUE TITLE
Advanced Search

### ISSUE THREAD

**Issue by pchainho** [reTHINK-project/dev-registry-domain#8]
Currently, searches for hyperty instances are limited to user. More advanced searches, like search an hyperty for a voice call with a certain user, imply the download and parse of Hyperty descriptors and object schemas from the catalogue.

This is limited by the current [Hyperty Instance data object model](https://github.com/reTHINK-project/dev-service-framework/tree/master/docs/datamodel/hyperty-registry#hyperty-instance)

Proposal:

To use the URL schemes to identify the data object types supported by each Hyperty Instance. Optionaly, the [Hyperty Resource type](https://github.com/reTHINK-project/dev-service-framework/tree/master/docs/datamodel/communication#hyperty-resource) can be used to enrich the search query criteria.

**Comment by emmelmann-fokus** [comment]
Hi Paulo,

could you specify what exactly you have in mind, e.g. giving an example?  I admit I don't fully understand your proposal.

**Comment by pchainho** [comment]
Hi Marc

the basic ideia is to use Data Objects URL schemes (<scheme>://<object-domain>/<object-identifier>) to identify its type. Eg from `call://example.com/dt3h5-f5gfgf` URL address, you could know this is the address of a voice call. In more complex situations where data objects are "richer" this may not be enough. For example, a collaborative communication where different types of resources can be shared among peers, say, eg, video, chat, file sharing, etc, besides the data object address you would also use the [Hyperty Resource type](https://github.com/reTHINK-project/dev-service-framework/tree/master/docs/datamodel/communication#hyperty-resource). For this example, the URL address of a communication Data Object instance would be something like:

 `comm://example.com/w6eg74-f764jkf` 

and a list of Hyperty Resource Types:

`["voice","video", "chat"]`

The Hyperty Instance that is handling this communication would be described in the domain registry with the list of supported URL schemes (ie Hyperty Data Object Types) and the list of supported Hyperty Resource Types eg:

```
{
	"schemes": ["comm"],
	"resource-types": ["voice", "video", "chat"]
}
```

In this way advanced searches for hyperty instances supporting voice call could be performed by just querying the domain registry on "schemes" and "resource types" ("comm", "call" schemes and "voice" resource types) and no interaction with the catalogue would be needed.

**Comment by emmelmann-fokus** [comment]
Hi Paulo,

thanks for providing the details.

**Comment by rjflp** [comment]
@pchainho Your proposal is feasible. It does improve search performance, but only in the situation when we already know which Domain Registry to query. If we only know the GUID, we will have to retrieve information on all the Service Providers being used by that user and contact each Domain Registry individually.

Is this information expected to be static? Could we add some of it to the Global Registry?

@fbeierle @sgoendoer What is your opinion?

**Comment by sbecot** [comment]
You don't ask me, but my opinion is that it is not really related to the global registry, which is dedicated to contact reachability. It would be more relevant in the search engine.
Another thing is that you can support a resource type but forbid it by your policy. eg: I have a video enabled hyperty, but currently my cam is off, or I simply don't want to broadcast. In this case the only one that can be aware of this is the domain registry.

**Comment by jmcrom** [comment]
@rjflp indeed this level of detail seems CSP-specific.

@pchainho I would think that _advanced search_ should be carried out by a discovery service.

**Comment by pchainho** [comment]
I've discussed this in more detail with @rjflp and we've agreed this should be supported by the domain registry

@jmcrom perhaps the title is misleading since this is actually pretty basic search that has to be supported from the beginning.

In order to support this kind of search I propose a minimal change in the [Hyperty Descriptor](https://github.com/reTHINK-project/dev-service-framework/blob/develop/docs/datamodel/core/hyperty-catalogue/readme.md#hyperty-descriptor) where the HypertyType has now two attributes (array of objects) and the [Hyperty Instance Registry](https://github.com/reTHINK-project/dev-service-framework/blob/develop/docs/datamodel/core/hyperty-registry/readme.md#hyperty-instance) will have also a new HypertyType attribute.

**Comment by acheambe** [comment]
@pchainho 
the HypertyType use to be represented by either the following HypertyTypes (Comunicator, Identity or context).
If i get you right, the HypertyType should now be represented as an array of HypertyResourceType. I can not find anywhere the specification of the HypertyResourceType. Can you give me a pointer pls? 
Or do you mean an array of HypertyTypes?

**Comment by rjflp** [comment]
Search for provided resources and schemes is already implemented in the Domain Registry, Connector and Runtime. Access to DataObject by name is also implemented. A release should be pushed today with these features.

Still to do:
- Search DataObject by hyperty reporter
- Search DataObject by its ID (url)

**Comment by antonroman** [comment]
Thanks for the update @rjflp 

regarding this point: _Access to DataObject by name is also implemented._  when you search by name what specific attirbute is being checked?: https://github.com/reTHINK-project/dev-service-framework/tree/master/docs/datamodel/core/hyperty-registry

In our application we needed to perform a search by "descriptor" to find a specific hyperty the user has registered (in our App each user has 3 instances of different hyperties in the Registry at the same time). Maybe implementing a search by descriptor could be helpful. @pchainho does it make sense or should we follow another approach? 

Thanks! and sorry If I'm doing a too obvious question.

**Comment by rjflp** [comment]
@antonroman You are right. The [documentation](https://github.com/reTHINK-project/dev-service-framework/tree/master/docs/datamodel/core/hyperty-registry) is not up-to-date. It should be updated to include the [SyncDataObject](https://github.com/reTHINK-project/dev-service-framework/tree/master/docs/datamodel/data-objects/data-synch). @pchainho can you do this, please?

@antonroman "Access to DataObject by name" was confusing. We enable search by URL and HypertyReporterURL (url field in [SyncDataObject](https://github.com/reTHINK-project/dev-service-framework/tree/master/docs/datamodel/data-objects/data-synch)).

Regarding the search by "descriptor" this is obviously possible, but should go through a "formal process". Please open a separate issue.

**Comment by pchainho** [comment]
The most updated in in the [dev branch](https://github.com/reTHINK-project/dev-service-framework/tree/develop/docs/datamodel/core/hyperty-registry)

Regarding the "search by name" this is really the purpose since all Objects has an attribute "name" that is registered in the Domain Registry.

**Comment by pchainho** [comment]
I'm reopening this issue since it seems "advanced" search of data objects is still not implemented

**Comment by nunofmn** [comment]
Implemented in release 0.3.0 740550d.
